File: py/medium/2225/2225.py
from collections import Counter
from itertools import chain
from typing import List


class Solution:
    def findWinners(self, matches: List[List[int]]) -> List[List[int]]:
        losers = map(lambda x: x[1], matches)
        counter_losers = Counter(losers)

        players = set(chain.from_iterable(matches))
        answer = [[], []]

        for player in players:
            if player not in counter_losers:
                answer[0].append(player)
            if player in counter_losers and counter_losers[player] == 1:
                answer[1].append(player)
        answer = [sorted(answer[0]), sorted(answer[1])]
        return answer


solution = Solution()
assert solution.findWinners(
    [[1, 3], [2, 3], [3, 6], [5, 6], [5, 7], [4, 5], [4, 8], [4, 9], [10, 4], [10, 9]]
) == [[1, 2, 10], [4, 5, 7, 8]]
assert solution.findWinners([[2, 3], [1, 3], [5, 4], [6, 4]]) == [[1, 2, 5, 6], []]

File: elixir/medium/2225/2225.exs
defmodule Solution do
  @spec find_winners(matches :: [[integer]]) :: [[integer]]
  def find_winners(matches) do
    players =
      List.flatten(matches)
      |> MapSet.new()

    losers_frequencies =
      matches
      |> Enum.map(fn [_, x] -> x end)
      |> Enum.frequencies()

    [0, 1]
    |> Enum.map(fn x -> losers_filter(players, losers_frequencies, x) end)
  end

  @spec losers_filter(players :: [integer], frequencies :: Map, n :: integer) :: [[integer]]
  def losers_filter(players, frequencies, n) do
    players
    |> Enum.filter(fn player ->
      if Map.get(frequencies, player, 0) == n, do: true, else: false
    end)
    |> Enum.sort()
  end
end

ExUnit.start()

defmodule AssertionTest do
  use ExUnit.Case, async: true

  test "test losers_filter 1" do
    assert Solution.losers_filter([1, 2, 3], %{1 => 1, 2 => 2, 3 => 3}, 1) == [1]
  end

  test "test losers_filter 2" do
    assert Solution.losers_filter([1, 2, 3], %{1 => 0, 2 => 0, 3 => 3}, 0) == [1, 2]
  end

  test "test_1" do
    assert Solution.find_winners([
             [1, 3],
             [2, 3],
             [3, 6],
             [5, 6],
             [5, 7],
             [4, 5],
             [4, 8],
             [4, 9],
             [10, 4],
             [10, 9]
           ]) == [[1, 2, 10], [4, 5, 7, 8]]
  end

  test "test_2" do
    assert Solution.find_winners([[2, 3], [1, 3], [5, 4], [6, 4]]) == [[1, 2, 5, 6], []]
  end
end
